dhis-2/dhis-services/dhis-service-validation/src/main/java/org/hisp/dhis/outlierdetection/service/MinMaxOutlierDetectionManager.java

```diff
@@ -0,0 +1,160 @@
+     *
+     * @param request the {@link OutlierDetectionRequest}.
+     * @return a list of {@link OutlierValue}.
+     */
+    public List<OutlierValue> getOutlierValues( OutlierDetectionRequest request )
+    {
+        final String ouPathClause = getOrgUnitPathClause( request.getOrgUnits() );
+
+        final String sql =
+            "select de.uid as de_uid, ou.uid as ou_uid, coc.uid as coc_uid, aoc.uid as aoc_uid, " +
+                "de.name as de_name, ou.name as ou_name, coc.name as coc_name, aoc.name as aoc_name, " +
+                "pe.startdate as pe_start_date, pt.name as pt_name, " +
+                "dv.value::double precision, " +
+                "least(abs(dv.value::double precision - mm.minimumvalue), abs(dv.value::double precision - mm.maximumvalue)) as bound_abs_dev, " +
+                "mm.minimumvalue as lower_bound, " +
+                "mm.maximumvalue as upper_bound " +
+            "from datavalue dv " +
+            "inner join dataelement de on dv.dataelementid = de.dataelementid " +
+            "inner join categoryoptioncombo coc on dv.categoryoptioncomboid = coc.categoryoptioncomboid " +
+            "inner join categoryoptioncombo aoc on dv.attributeoptioncomboid = aoc.categoryoptioncomboid " +
+            "inner join period pe on dv.periodid = pe.periodid " +
+            "inner join periodtype pt on pe.periodtypeid = pt.periodtypeid " +
+            "inner join organisationunit ou on dv.sourceid = ou.organisationunitid " +
+            // Min-max data element join
+            "inner join minmaxdataelement mm on (dv.dataelementid = mm.dataelementid " +
+                "and dv.sourceid = mm.sourceid and dv.categoryoptioncomboid = mm.categoryoptioncomboid) " +
+            "where dv.dataelementid in (:data_element_ids) " +
+            "and pe.startdate >= :start_date " +
+            "and pe.enddate <= :end_date " +
+            "and " + ouPathClause + " " +
+            "and dv.deleted is false " +
+            // Filter for values outside the min-max range
+            "and (dv.value::double precision < mm.minimumvalue or dv.value::double precision > mm.maximumvalue) " +
+            // Order and limit
+            "order by bound_abs_dev desc " +
+            "limit :max_results;";
+
+        final SqlParameterSource params = new MapSqlParameterSource()
+            .addValue( "data_element_ids", request.getDataElementIds() )
+            .addValue( "start_date", request.getStartDate() )
+            .addValue( "end_date", request.getEndDate() )
+            .addValue( "max_results", request.getMaxResults() );
+
+        final Calendar calendar = PeriodType.getCalendar();
+
+        try
+        {
+            return jdbcTemplate.query( sql, params, getRowMapper( calendar ) );
+        }
+        catch ( DataIntegrityViolationException ex )
+        {
+            // Casting non-numeric data to double, catching exception is faster than filtering
```

Comment:
Is this an input error?

Reply:
The background here is that we store data as text/varchar in the database for the `datavalue` table, as we use the same column for all types of values (numeric, bool, text, etc). So there is the need to cast to double for when we need to do numeric operations. We do check the _value type_ of the data element before executing this query, so if there are no integrity issues in the database, this will not happen. In the case of integrity issues we might get a cast exception here trying to cast text to double. The fix would be to use `regexp_match` to filter for numeric values first, but this is slow and makes the query x3 slower, so it is not worth the trade-off and it makes more sense to just catch the exception in the rare case it happens.